src/GapsInCareHelpers.ts

```diff
@@ -355,16 +368,96 @@ export function generateGapsInCareBundle(
+ * @param detailedResult result from calculation to look up clause values
+ * @return mapped list of queries with near miss info if relevant
  */
-export function calculateNearMisses(retrieves: DataTypeQuery[], improvementNotation: string): DataTypeQuery[] {
+export function calculateNearMisses(
+  retrieves: DataTypeQuery[],
+  improvementNotation: string,
+  detailedResult?: DetailedPopulationGroupResult
+): DataTypeQuery[] {
   return retrieves.map(r => {
-    let isNearMiss;
+    let nearMissInfo: NearMissInfo;
     if (improvementNotation === ImprovementNotation.POSITIVE) {
-      isNearMiss = r.retrieveHasResult && !r.parentQueryHasResult;
+      nearMissInfo = {
+        isNearMiss: r.retrieveHasResult === true && r.parentQueryHasResult === false,
+        reasonCodes: []
+      };
+
+      if (nearMissInfo.isNearMiss && r.queryInfo && detailedResult?.clauseResults) {
+        const flattenedFilters = flattenFilters(r.queryInfo.filter);
+
+        flattenedFilters.forEach(f => {
+          // Separate interval handling for 'during' filters
+          if (f.type === 'during') {
+            const duringFilter = f as DuringFilter;
+            const resources = detailedResult.clauseResults?.find(
+              cr => cr.libraryName === r.libraryName && cr.localId === r.retrieveLocalId
+            );
+
+            if (duringFilter.valuePeriod.interval && resources) {
+              const path = duringFilter.attribute.split('.');
+
+              // Access desired property of FHIRObject
+              resources.raw.forEach((r: any) => {
+                let desiredAttr = r;
+                path.forEach(key => {
+                  if (desiredAttr.value?.isDateTime) {
+                    return;
+                  }
+
+                  desiredAttr = desiredAttr[key];
+                });
+
+                // Use DateOutOfRange code if data point is outside of the desired interval
+                const isAttrContainedInInterval = duringFilter.valuePeriod.interval?.contains(desiredAttr.value);
+
+                if (isAttrContainedInInterval === false) {
+                  nearMissInfo.reasonCodes?.push(CareGapReasonCode.DATEOUTOFRANGE);
+                }
+              });
+            }
+          } else {
+            // TODO: This logic is not perfect, and can be corrupted by multiple resources spanning truthy values for all filters
+            // For non-during filters, look up clause result by localId
+            const clauseResult = detailedResult.clauseResults?.find(
+              cr => cr.libraryName === r.libraryName && cr.localId === f.localId
+            );
+
+            // False clause means this specific filter was falsy
+            if (clauseResult && clauseResult.final === FinalResult.FALSE) {
+              const code = getGapReasonCode(f);
+              if (code !== null) {
+                nearMissInfo.reasonCodes?.push(code);
```

Comment:
Same suggestion here as above

Reply:
Details hashed out in slack. But I went ahead and made reasonCodes a required prop for near misses, which is fine. Is empty array in the bad case.